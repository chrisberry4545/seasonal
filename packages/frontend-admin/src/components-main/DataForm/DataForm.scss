.c-data-form__field {
  display: block;
}
