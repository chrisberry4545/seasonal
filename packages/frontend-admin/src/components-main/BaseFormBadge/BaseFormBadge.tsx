import React, { FC } from 'react';
import {
  IGetAuthorizedBackendDataProps
} from '../GetAuthorizedBackendData/GetAuthorizedBackendData';
import { IBadge } from '@chrisb-dev/seasonal-shared-models';
import { IDataFormConfigProps, DataForm } from '../DataForm/DataForm';
import {
  hexColorValidation,
  requiredValidation
} from '@chrisb-dev/seasonal-shared-frontend-components';

type IBadgeFormConfigProps = IDataFormConfigProps<IBadge>;

const badgeFormConfig: IBadgeFormConfigProps = {
  name: {
    type: 'text',
    validation: [requiredValidation]
  },

  color: {
    type: 'text',
    validation: [requiredValidation, hexColorValidation]
  }
};

export const BaseFormBadge: FC<IGetAuthorizedBackendDataProps<IBadge>> = ({
  items,
  updateMethod,
  buttonText
}) => (
  <DataForm item={items}
    sendData={updateMethod}
    formConfig={badgeFormConfig}
    buttonText={buttonText} />
);
