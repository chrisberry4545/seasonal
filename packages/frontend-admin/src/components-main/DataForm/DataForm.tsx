import React, { useState } from 'react';
import {
  Checkbox,
  Input,
  ISelectOption,
  Multiselect,
  Select,
  IValidation,
  LoadingSpinner,
  PrimaryButton,
  TextMedium
} from '@chrisb-dev/seasonal-shared-frontend-components';
import './DataForm.scss';
import { FORM_BUTTON_TEXT } from '../../consts';

export interface IFormField {
  name?: string;
  options?: ISelectOption[];
  type: 'text' | 'number' | 'password' | 'checkbox' | 'select' | 'multiselect';
  validation?: IValidation[];
}
export type IDataFormConfigProps<T> = { [key in keyof T & string]?: IFormField };
export interface IDateFormProps<T> {
  item: Partial<T>;
  sendData?: (data: T) => Promise<T | void>;
  formConfig: IDataFormConfigProps<T> | null;
  processItem?: (
    item: Partial<T>,
    previousItem: Partial<T> | null
  ) => Partial<T>;
  buttonText?: string;
}

export function DataForm<T>({
  item,
  sendData,
  formConfig,
  processItem,
  buttonText = FORM_BUTTON_TEXT.UPDATE
}: IDateFormProps<T>) {
  const [itemState, setItemState] = useState<Partial<T>>({
    ...item
  });
  const [errorState, setErrorState] = useState<string | null>(null);
  const [validationState, setValidationState] = useState<
    { [key in keyof T]?: string[] }
  >({});
  const [isLoadingState, setIsLoadingState] = useState<boolean>(false);

  const updateItem = (
    newItem: Partial<T>
  ) => {
    if (processItem) {
      const processedItem = processItem(newItem, itemState);
      setItemState(processedItem);
    } else {
      setItemState(newItem);
    }
  };

  const submit = async () => {
    try {
      setIsLoadingState(true);
      const updatedItem = await sendData!(itemState as T);
      if (updatedItem) {
        updateItem(updatedItem);
        setIsLoadingState(false);
      }
    } catch (error) {
      setIsLoadingState(false);
      setErrorState(error.message);
    }
  };

  const updateField = (
    name: keyof T,
    value: string | string[] | number | boolean,
    validation: IValidation[] | undefined
  ) => {
    if (validation) {
      const errors = validation.map((validationFunction) =>
        validationFunction(value)
      );
      setValidationState({
        ...validationState,
        name: errors
      });
    }
    const newItem = {
      ...itemState,
      [name]: value
    };
    updateItem(newItem);
  };

  return (
    <div className='c-data-form'>
      <div>
        {
          formConfig && Object.entries(formConfig).map(([key, formField]) => {
            const { options, type, validation } = formField as IFormField;
            const prop = key as keyof T;
            const value = itemState[prop] as any;
            const placeholder = key[0].toUpperCase()
              + key.replace( /([A-Z])/g, ' $1').slice(1);
            const inputs = {
              className: 'c-data-form__input',
              onChange: (
                changedValue: number | string | string[] | boolean
              ) => updateField(prop, changedValue, validation),
              options,
              placeholder,
              type,
              value
            };
            const validationErrors = validationState[prop];
            return (
              <label key={key} className='c-data-form__field'>
                <TextMedium className='c-data-form__label'>{placeholder}</TextMedium>
                {
                  (() => {
                    switch (type) {
                      case 'number':
                      case 'text':
                      case 'password':
                        return <Input {...{
                          ...inputs,
                          type: inputs.type as 'text' | 'number' }
                        } />;
                      case 'checkbox':
                        return <Checkbox {...inputs} />;
                      case 'multiselect':
                        return <Multiselect {...{
                          ...inputs,
                          options: inputs.options!
                        } as any} />;
                      case 'select':
                        return <Select {...{
                          ...inputs,
                          options: inputs.options!
                        }} />;
                    }
                  })()
                }
                <div>
                  { validationErrors && validationErrors[0] }
                </div>
              </label>
            );
          })
        }
      </div>
      <PrimaryButton className='c-data-form__submit-btn' onClick={submit}>
        {buttonText}
      </PrimaryButton>
      {
        errorState ? <div>{errorState}</div> : null
      }
      {
        isLoadingState
        && <div className='c-data-form__loading-spinner'>
            <LoadingSpinner />
          </div>
      }
    </div>
  );
}
