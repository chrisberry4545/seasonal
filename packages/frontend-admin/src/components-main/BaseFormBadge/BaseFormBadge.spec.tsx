import React from 'react';
import { BaseFormBadge } from './BaseFormBadge';
import { shallow, ShallowWrapper } from 'enzyme';
import { IBadge } from '@chrisb-dev/seasonal-shared-models';
import { DataForm } from '../DataForm/DataForm';

jest.mock('../DataForm/DataForm', () => ({
  DataForm: () => 'DataForm'
}));

describe('<BaseFormBadge />', () => {
  let wrapper: ShallowWrapper;
  const item = {} as IBadge;
  let mockUpdateMethod: jest.Mock;

  beforeEach(() => {
    mockUpdateMethod = jest.fn();
    wrapper = shallow(
      <BaseFormBadge
        items={item}
        updateMethod={mockUpdateMethod}
        buttonText='Edit'
        />
    );
  });

  test('renders correctly', () => expect(wrapper).toMatchSnapshot());

  test('can call update method', () => {
    const sendData = wrapper.find(DataForm).props().sendData;
    if (sendData) {
      sendData(item);
    }
    expect(mockUpdateMethod).toHaveBeenCalledWith(item);
  });

});
