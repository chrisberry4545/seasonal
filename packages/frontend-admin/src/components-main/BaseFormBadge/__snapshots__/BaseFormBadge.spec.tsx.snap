// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseFormBadge /> renders correctly 1`] = `
<DataForm
  buttonText="Edit"
  formConfig={
    Object {
      "name": Object {
        "type": "text",
        "validation": Array [
          [Function],
        ],
      },
    }
  }
  item={Object {}}
  sendData={[MockFunction]}
/>
`;
